libraries/AP_RTC/AP_RTC.cpp

```diff
@@ -0,0 +1,125 @@
+#include "AP_RTC.h"
+#include <AP_HAL/AP_HAL.h>
+#include <GCS_MAVLink/GCS.h>
+
+extern const AP_HAL::HAL& hal;
+
+uint64_t AP_RTC::get() const
+{
+    if (!_is_set) {
+        return 0U;
+    }
+    return rtc_shift + AP_HAL::micros64();
+}
+
+void AP_RTC::set(uint64_t time_utc_usec) {
+    if (_is_set) {
+        // you can only set once
+        return;
+    }
+    if (time_utc_usec < AP_HAL::micros64()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "RTC: Invalid UTC time (%u)", rtc_shift);
+        return;
+    }
+    rtc_shift = time_utc_usec - AP_HAL::micros64();
```

Comment:
I wonder if this shift should be given by the caller, which should have it more accurate.